contracts/multisig-prover/src/encoding/mvx/mod.rs

```diff
@@ -0,0 +1,438 @@
+    type Error = ContractError;
+
+    fn try_from(msg: &RouterMessage) -> Result<Self, Self::Error> {
+        let map_addr_err = |_| ContractError::InvalidMessage {
+            reason: format!(
+                "destination_address is not a valid MVX address: {}",
+                msg.destination_address.as_str()
+            ),
+        };
+
+        let (_, data, _) =
+            bech32::decode(&msg.destination_address.as_str()).map_err(map_addr_err)?;
+        let addr_vec = Vec::<u8>::from_base32(&data).map_err(map_addr_err)?;
+        let contract_address =
+            <[u8; 32]>::try_from(addr_vec).map_err(|_| ContractError::InvalidMessage {
+                reason: format!(
+                    "destination_address is not a valid MVX address: {}",
+                    msg.destination_address.as_str()
+                ),
+            })?;
+
+        Ok(Message {
+            source_chain: msg.cc_id.chain.to_string(),
+            message_id: msg.cc_id.id.to_string(),
+            source_address: msg.source_address.to_string(),
+            contract_address,
+            payload_hash: msg.payload_hash,
+        })
+    }
+}
+
+fn uint256_to_compact_vec(value: Uint256) -> Vec<u8> {
+    if value.is_zero() {
+        return Vec::new();
+    }
+
+    let bytes = value.to_be_bytes();
+    let mut slice_from = 0;
+    for (i, byte) in bytes.iter().enumerate() {
+        if *byte != 0 {
+            slice_from = i;
+            break;
+        }
+    }
+
+    bytes[slice_from..].to_vec()
+}
+
+pub fn ed25519_key(pub_key: &PublicKey) -> Result<[u8; 32], ContractError> {
+    match pub_key {
+        PublicKey::Ed25519(ed25519_key) => {
+            return Ok(<[u8; 32]>::try_from(ed25519_key.as_ref())
+                .expect("couldn't convert pubkey to ed25519 public key"));
+        }
+        _ => {
+            return Err(ContractError::InvalidPublicKey {
+                reason: "Public key is not ed25519".into(),
+            })
+        }
+    }
+}
+
+pub fn payload_hash_to_sign(
+    domain_separator: &Hash,
+    signer: &VerifierSet,
+    payload: &Payload,
+) -> Result<Hash, ContractError> {
+    let signer_hash = WeightedSigners::from(signer).hash();
+    let data_hash = Keccak256::digest(encode(payload)?);
+
+    let unsigned = [
+        "\x19MultiversX Signed Message:\n".as_bytes(),
+        domain_separator,
+        signer_hash.as_slice(),
+        data_hash.as_slice(),
+    ]
+    .concat();
+
+    Ok(Keccak256::digest(unsigned).into())
+}
+
+fn encode(payload: &Payload) -> Result<Vec<u8>, ContractError> {
+    match payload {
+        Payload::Messages(messages) => {
+            let messages: Vec<_> = messages
+                .iter()
+                .map(Message::try_from)
+                .collect::<Result<_, _>>()?;
+
+            let mut result = vec![0u8];
+            let mut messages =
+                top_encode_to_vec_u8(&messages).expect("couldn't serialize messages as mvx");

```

Comment:
Hm, the only way to do this I think is to use `into_iter().chain()`. That works, but I feel like using this append is more efficient.
The problem is that `0u8` needs to be appended to the beginning of this vec (because of Mvx encoding particularities), and can not be done directly through `top_encode_to_vec_u8` unfortunately...